### PR TITLE
admin: fix field mask parameter

### PR DESCRIPTION
**Pull request by paul-r-gall**
Signed-off-by: Paul Gallagher <pgal@google.com>

Commit Message: admin: fix field mask parameter
Additional Description: PR #16979 caused a regression where providing field masks without providing a resource did not work. This change makes it so that providing a bad field mask fails silently when a resource is not provided.
Risk Level: low
Testing: unit test added to detect regression so that it will be caught in the future.

### CODE REVIEW

**Comment by paul-r-gall** [comment]
@jmarantz could you take a look at this fix?

File: source/server/admin/config_dump_handler.cc [hunk 1]
```diff
@@ -241,11 +241,11 @@ ConfigDumpHandler::addAllConfigToDump(envoy::admin::v3::ConfigDump& dump,
...
       // We don't use trimMessage() above here since masks don't support
-      // indexing through repeated fields.
+      // indexing through repeated fields. We don't return error on failure
+      // because different callback return types will have different valid
+      // field masks.
       if (!checkFieldMaskAndTrimMessage(field_mask, *message)) {
-        return absl::optional<std::pair<Http::Code, std::string>>{std::make_pair(
-            Http::Code::BadRequest, absl::StrCat("FieldMask ", field_mask.DebugString(),
-                                                 " could not be successfully used."))};
+        continue;
```
Comment (jmarantz) [inline source/server/admin/config_dump_handler.cc:248]: WDYT about just including in the output text the error message that we had before. IIUC you were taking a case that had an instructive error message before about what was wrong, and then making it just return an empty string.

Maybe collect the names of the fields that had invalid masks and return them in one error message?

We might also return an http error when none of the clusters had a useful mask. WDYT?

Comment (paul-r-gall) [reply source/server/admin/config_dump_handler.cc:248]: I'm considering the case in the test, when someone just supplies `mask=bootstrap`. Intended behavior there is just that the bootstrap config is returned, with no error message. However, that mask *is* invalid for the majority of protos in the dump (which is why this was failing before).

I think that returning an HTTP error when no clusters had a useful mask is a good idea.

Comment (jmarantz) [reply source/server/admin/config_dump_handler.cc:248]: sgtm.

**Review by jmarantz — COMMENTED** [review]
/wait

File: source/server/admin/config_dump_handler.cc [hunk 2]
```diff
@@ -282,6 +282,11 @@ absl::optional<std::pair<Http::Code, std::string>> ConfigDumpHandler::addAllConf
     auto* config = dump.add_configs();
     config->PackFrom(*message);
   }
+  if (dump.configs().empty() && mask.has_value()) {
```
Comment (jmarantz) [inline source/server/admin/config_dump_handler.cc:285]: can you just use `config` here rather than calling `dump->configs()`?

**Review by jmarantz — APPROVED** [review]
Looks great with one minor nit.

/assign-from @envoyproxy/senior-maintainers

File: source/server/admin/config_dump_handler.cc (hunk 2 above)
Comment (paul-r-gall) [reply source/server/admin/config_dump_handler.cc:285]: No, `config` is out of scope at this point.

Comment (jmarantz) [reply source/server/admin/config_dump_handler.cc:285]: indeed :)

**Pull request merged**